apps/builder/app/builder/features/ai/copywriter.tsx

```diff
@@ -0,0 +1,187 @@
+import { z } from "zod";
+import store from "immerhin";
+import untruncateJson from "untruncate-json";
+import { copywriter, requestStream } from "@webstudio-is/ai";
+import { restAiCopy } from "~/shared/router-utils";
+import {
+  Box,
+  Button,
+  InputField,
+  Label,
+  Text,
+} from "@webstudio-is/design-system";
+import {
+  instancesStore,
+  projectStore,
+  selectedInstanceStore,
+} from "~/shared/nano-states";
+import { useStore } from "@nanostores/react";
+import { collectTextInstances } from "node_modules/@webstudio-is/ai/src/chains/copywriter";
+import { useMemo, useRef, useState } from "react";
+
+const patchTextInstance = (textInstance: copywriter.TextInstance) => {
+  store.createTransaction([instancesStore], (instances) => {
+    const currentInstance = instances.get(textInstance.instanceId);
+
+    if (
+      currentInstance === undefined ||
+      currentInstance.children.length === 0
+    ) {
+      return;
+    }
+
+    // Instances can have a number of text child nodes without interleaving components.
+    // When this is the case we treat the child nodes as a single text node,
+    // otherwise the AI would generate children.length chunks of separate text.
+    // We can identify this case of "joint" text instances when the index is -1.
+    const replaceAll = textInstance.index === -1;
+    if (replaceAll) {
+      if (currentInstance.children.every((child) => child.type === "text")) {
+        currentInstance.children = [{ type: "text", value: textInstance.text }];
+      }
+      return;
+    }
+
+    if (currentInstance.children[textInstance.index].type === "text") {
+      currentInstance.children[textInstance.index].value = textInstance.text;
+      instances.set(textInstance.instanceId, currentInstance);
```

Comment:
is this necessary or mutating `currentInstance` will just work? I am asking because I realized that at line 40 I don't do an `instance.set` after the update